contract/models/contract.py

```diff
@@ -0,0 +1,260 @@
+        if product.description_sale:
+            name += '\n' + product.description_sale
+        vals['name'] = name
+
+        vals['price_unit'] = product.price
+        self.update(vals)
+        return {'domain': domain}
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = 'account.analytic.account'
+
+    @api.model
+    def _default_journal(self):
+        company_id = self.env.context.get(
+            'company_id', self.env.user.company_id.id)
+        domain = [
+            ('type', '=', 'sale'),
+            ('company_id', '=', company_id)]
+        return self.env['account.journal'].search(domain, limit=1)
+
+    pricelist_id = fields.Many2one(
+        comodel_name='product.pricelist',
+        string='Pricelist')
+    date_start = fields.Date(default=fields.Date.context_today)
+    recurring_invoice_line_ids = fields.One2many(
+        comodel_name='account.analytic.invoice.line',
+        inverse_name='analytic_account_id',
+        string='Invoice Lines')
+    recurring_invoices = fields.Boolean(
+        string='Generate recurring invoices automatically')
+    recurring_rule_type = fields.Selection(
+        [('daily', 'Day(s)'),
+         ('weekly', 'Week(s)'),
+         ('monthly', 'Month(s)'),
+         ('yearly', 'Year(s)'),
+         ],
+        default='monthly',
+        string='Recurrency',
+        help="Invoice automatically repeat at specified interval")
+    recurring_interval = fields.Integer(
+        default=1,
+        string='Repeat Every',
+        help="Repeat every (Days/Week/Month/Year)")
+    recurring_next_date = fields.Date(
+        default=fields.Date.context_today,
+        string='Date of Next Invoice')
+    journal_id = fields.Many2one(
+        'account.journal',
+        string='Journal',
+        default=_default_journal,
+        domain="[('type', '=', 'sale'),('company_id', '=', company_id)]")
+
+    @api.multi
+    def copy(self, default=None):
+        default = dict(default or {})
+        # Reset next invoice date
+        default.update(
+            recurring_next_date=self._defaults['recurring_next_date'](self)
+        )
+        return super(AccountAnalyticAccount, self).copy(default=default)
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        self.pricelist_id = self.partner_id.property_product_pricelist.id
+
+    @api.onchange('recurring_invoices')
+    def _onchange_recurring_invoices(self):
+        if self.date_start and self.recurring_invoices:
+            self.recurring_next_date = self.date_start
+
+    @api.model
+    def _insert_markers(self, line, date_start, next_date, date_format):
+        line = line.replace('#START#', date_start.strftime(date_format))
+        date_end = next_date - relativedelta(days=1)
+        line = line.replace('#END#', date_end.strftime(date_format))
+        return line
+
+    @api.model
+    def _prepare_invoice_line(self, line, invoice_id):
+        product = line.product_id
```

Comment:
Here you have to use the same technique as in https://github.com/OCA/commission/pull/67/files#diff-73f037396680f271cd1a6b014c9935ecR86 to use the invoice line `onchange_product_id`and this way, you won't need to make a lot of glue modules that inject their values on it.